pkg/tbtc/deduplicator.go

```diff
@@ -90,3 +99,37 @@ func (d *deduplicator) notifyDKGResultSubmitted(
 	// proceed with the execution.
 	return false
 }
+
+// notifyDepositSweepProposalSubmitted notifies the client wants to start some
+// actions upon the deposit sweep proposal submission. It returns boolean
+// indicating whether the client should proceed with the actions or ignore the
+// event as a duplicate.
+func (d *deduplicator) notifyDepositSweepProposalSubmitted(
+	newProposal *DepositSweepProposal,
+) bool {
+	d.depositSweepProposalCache.Sweep()
+
+	var buffer bytes.Buffer
+	buffer.Write(newProposal.WalletPubKeyHash[:])
+	for _, depositKey := range newProposal.DepositsKeys {
+		buffer.Write(depositKey.FundingTxHash[:])
+		fundingOutputIndex := make([]byte, 4)
+		binary.BigEndian.PutUint32(fundingOutputIndex, depositKey.FundingOutputIndex)
+		buffer.Write(fundingOutputIndex)
+	}
+	buffer.Write(newProposal.SweepTxFee.Bytes())
+
+	bufferSha256 := sha256.Sum256(buffer.Bytes())
+	cacheKey := hex.EncodeToString(bufferSha256[:])
```

Comment:
It makes sense that the deduplicator is doing only what it is supposed to be doing: deduplicating an event, without going deeper and not trying to take the responsibility of wallet locking.

My only comment to this code is that the hashing logic should live next to `type DepositSweepProposal struct`. I think we should have some sort of `HashCode()` function (like in JVM world) defined next to the struct and use this function here. This way, if we add one more field to the struct, it will be easier to notice that the hashing must be updated. Also, there is a chance we'll reuse the `HashCode()` in other places.

I think all the tests in `deduplicator_test.go` should stay as-is, I'd just add some tests for `HashCode()`.

Reply:
I was thinking about it but that depends on how we look at the "hash code" function. I'm rather looking at it as an internal deduplication-specific thing, hence why it is placed here. My reasoning: the deduplicator must decide what is and what is not a duplicate, hence it must build the cache key accordingly. Having a general "hash code" function exposed by `DepositSweepProposal` may be dangerous as their connection with the deduplicator will be not obvious and we can mess up things by adding/removing fields.

Reply:
See 47f9f482